megatron/data/data_utils.py

```diff
@@ -480,6 +480,97 @@ def build_train_valid_test_data_iterators(neox_args):
     return train_data_iterator, valid_data_iterator, test_data_iterator
 
 
+def build_validation_iterator(neox_args):
+    """XXX"""
+
+    valid_dataloader = None
+
+    print_rank_0("> building validation ...")
+
+    # Ensure only the first/last pipeline stages have data loaders
+    if neox_args.is_pipe_parallel:
+        is_first_stage = mpu.get_pipe_parallel_rank() == 0
+        is_last_stage = (
+            mpu.get_pipe_parallel_rank() == mpu.get_pipe_parallel_world_size() - 1
+        )
+        pipe_load = is_first_stage or is_last_stage
+    else:
+        pipe_load = True
+
+    # Data loader only on rank 0 of each model parallel group.
+    if mpu.get_model_parallel_rank() == 0 and pipe_load:
+        # Number of train/valid/test samples.
+        train_iters = neox_args.train_iters
+        eval_iters = (train_iters // neox_args.eval_interval + 1) * neox_args.eval_iters
+        test_iters = neox_args.eval_iters
+        num_samples = eval_iters * neox_args.train_batch_size
+            
+        valid_weights, valid_num_samples = get_normalized_weights_and_num_samples(
+            neox_args.valid_data_weights, num_samples
+        )
+
+        # build individual datasets
+        _, valid_datasets, _ = build_weighted_datasets(
+            neox_args,
+            0,
+            valid_num_samples,
+            0,
+            0,
+            valid_weights,
+            0,
+            build_index_mappings=not neox_args.weight_by_num_documents,
+        )
+
+        # if neox_args.weight_by_num_documents: # Not supported for now
+
+        if valid_datasets:
+            valid_ds = BlendableDataset(valid_datasets, valid_weights)
```

Comment:
Why do you use BlendableDataset here instead of build_train_valid_test_datasets() as was done here https://github.com/EleutherAI/gpt-neox/blob/649c30971d45326f140516ee7e333cdcef01062c/megatron/data/data_utils.py#LL400C43-L400C74?

Not necessarily something wrong just wondering.

Reply:
Because we have already split the dataset